src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -1745,6 +1745,20 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
                 );
                 throw SemanticAbort();
             }
+            if((ASRUtils::extract_kind_from_ttype_t(left_type) != ASRUtils::extract_kind_from_ttype_t(right_type)) || 
+                ((ASRUtils::is_logical(*left_type) && !ASRUtils::is_logical(*right_type))) ||
+                ((!ASRUtils::is_logical(*left_type) && ASRUtils::is_logical(*right_type)))){
+                std::string ltype = ASRUtils::type_to_str_python(ASRUtils::expr_type(left));
+                std::string rtype = ASRUtils::type_to_str_python(ASRUtils::expr_type(right));
+                diag.add(diag::Diagnostic(
+                    "Type mismatch in binary operator; the types must be compatible",
+                    diag::Level::Error, diag::Stage::Semantic, {
+                        diag::Label("type mismatch (" + ltype + " and " + rtype + ")",
+                                {left->base.loc, right->base.loc})
+                    })
+                );
+                throw SemanticAbort();
+            }
```

Comment:
Can we use the following logic here as well?

https://github.com/lcompilers/lpython/blob/57b497137b48eb0adb01e6f2b8dd6f3873b82c32/src/lpython/semantics/python_ast_to_asr.cpp#L1995-L2008

Reply:
Done. Thanks for pointing it out!